1. Basic Web Development/frontend/src/main.ts

```diff
@@ -0,0 +1,419 @@
+
+    const messageJSON = JSON.stringify({ index: index, type: type.replace('Edit', ''), text: text });
+
+    fetch(API_URL + '/pref/update', { method: 'PUT', headers: HEADERS, body: messageJSON })
+        .then(response => response.json())
+        .catch((error) => {
+            console.error('Error:', error);
+            permissionsRoutine();
+        });
+
+    createPermissionsForm(form);
+}
+
+function handlePopUp(lock: HTMLInputElement, lock2: HTMLInputElement, popUp: Element) {
+    lock.addEventListener('click', (event) => {
+        this.ISLOCKED = lock.checked;
+        switchLock(lock2, popUp);
+    });
+
+    lock2.addEventListener('click', (event) => {
+        this.ISLOCKED = lock2.checked;
+        switchLock(lock, popUp);
+    });
+
+    if (this.ISLOCKED) {
+        popUp.classList.add('showing');
+        popUp.classList.remove('hidden');
+    }
+
+    popUp.addEventListener('click', () => {
+        if (!this.ISLOCKED) {
+            popUp.classList.add('showing');
+            popUp.classList.remove('hidden');
+        }
+    });
+
+    popUp.addEventListener('mouseleave', () => {
+        console.log(this.ISLOCKED);
+        if (!this.ISLOCKED) {
+            popUp.classList.add('hidden');
+            popUp.classList.remove('showing');
+        }
+    });
+}
+
+function switchLock(lock: HTMLInputElement, popUp: Element) {
+    popUp.classList.remove('hidden');
+    popUp.classList.remove('showing');
+
+    if (this.ISLOCKED) {
+        popUp.classList.add('showing');
+        lock.checked = true;
+    } else {
+        popUp.classList.add('hidden');
+        lock.checked = false;
+    }
+}
+
+function changeBoxesState(boxes: NodeListOf<HTMLInputElement>, state: boolean) {
+    boxes.forEach(box => {
+        box.checked = state;
+    });
+}
+
+function getRandomID() {
+    let used: number[] = [];
+
+    function getNumber() {
+        let randomNo = Math.floor(Math.random() * 1000);
+        while (used.includes(randomNo)) {
+            randomNo = Math.floor(Math.random() * 1000);
+        }
+
+        used.push(randomNo);
+        return randomNo;
+    }
+    return getNumber;
+}
+
+function removeChildren(el: HTMLElement) {
+    let child = el.lastElementChild;
+    while (child) {
+        el.removeChild(child);
+        child = el.lastElementChild;
+    }
+}
+
+function toggleLoad(isLoading: boolean) {
+    if(isLoading) {
+        this.document.getElementById("permissions-list").classList.add("no-display");
+        this.document.getElementById("spinner").classList.remove("no-display");
+    } else {
+        this.document.getElementById("permissions-list").classList.remove("no-display");
+        this.document.getElementById("spinner").classList.add("no-display");
+    }
+}
+
+class Permission {
+    id: string;
+
+    constructor(public text: string, public type: string) {
+        this.type = type;
+        this.text = text;
+        this.id = text.split(' ')[0].toLowerCase() + randomGenerator();
+    }
```

Comment:
```suggestion
class Permission {
    id: string;
    type: PermTypes;
    
    constructor(public text: string, public type: string) {
        this.type = type;
        this.text = text;
        this.id = text.split(' ')[0].toLowerCase() + randomGenerator();
    }
```

Same thing as for the backend - use enums to prevent misspelling. Also id assignment should be delegated to backend.